### PR TITLE
api,server: using `ontology.Resource` message in `Evidence.resource`

### PR DESCRIPTION
**Pull request by oxisto**
Instead of an untested `protobuf.Any` we are now using the `ontology.Resource` message, which in turn has a one-of field to populate the `Evidence.resource` field.

This has the advantage that now a proper OpenAPI description is now generated for this field.

### CODE REVIEW

File: api/evidence/evidence.go [hunk 1]
```diff
@@ -41,40 +42,37 @@ func (req *StoreEvidenceRequest) GetPayload() proto.Message {
...
-	m, err = ev.Resource.UnmarshalNew()
-	if err != nil {
+
+	if ev.Resource == nil || ev.Resource.Type == nil {
 		return nil
 	}
 
-	resource, ok = m.(ontology.IsResource)
+	// A little bit of dark Go magic
+	typ := reflect.ValueOf(ev.Resource.Type).Elem()
```
Comment (lebogg) [inline api/evidence/evidence.go:54]: I saw that this `.Elem()` function could panic. Therefore I took a closer look and added some tests. It's good due to the nil checks above.

File: api/evidence/evidence.go [hunk 2]
```diff
@@ -41,40 +42,37 @@ func (req *StoreEvidenceRequest) GetPayload() proto.Message {
...
-	// Try to extract the resource out of the evidence
-	m, err = ev.Resource.UnmarshalNew()
-	if err != nil {
+
+	if ev.Resource == nil || ev.Resource.Type == nil {
 		return ""
 	}
 
-	resource, ok := m.(ontology.IsResource)
+	typ := reflect.ValueOf(ev.Resource.Type).Elem()
```
Comment (lebogg) [inline api/evidence/evidence.go:74]: I saw that this `.Elem()` function could panic. Therefore I took a closer look and added some tests and also a nil check like you did above

File: api/ontology/ontology.go [hunk 3]
```diff
@@ -2,10 +2,11 @@ package ontology
 
 import (
 	"encoding/json"
-	"google.golang.org/protobuf/reflect/protoreflect"
 	"slices"
 	"strings"
 
+	"google.golang.org/protobuf/reflect/protoreflect"
+
```
Comment (lebogg) [inline api/ontology/ontology.go:9]:
```suggestion
```

Comment (lebogg) [reply api/ontology/ontology.go:9]: empty line

File: api/ontology/ontology.go [hunk 4]
```diff
@@ -142,3 +143,30 @@ func ResourceIDs(r []IsResource) []string {
 
 	return a
 }
+
+// ProtoResource converrts a [IsResource] to a [Resource]
```
Comment (lebogg) [inline api/ontology/ontology.go:147]:
```suggestion
// ProtoResource converts a [IsResource] to a [Resource]
```

File: api/ontology/ontology.go [hunk 5]
```diff
@@ -142,3 +143,30 @@ func ResourceIDs(r []IsResource) []string {
...
+func ProtoResource(resource IsResource) *Resource {
+	var (
+		r  Resource
+		m  protoreflect.Message         = r.ProtoReflect()
+		od protoreflect.OneofDescriptor = m.Descriptor().Oneofs().ByName("type")
```
Comment (lebogg) [inline api/ontology/ontology.go:152]:
```suggestion
		m  = r.ProtoReflect()
		od = m.Descriptor().Oneofs().ByName("type")
```

Comment (lebogg) [reply api/ontology/ontology.go:152]: just a suggestion

Comment (lebogg) [reply api/ontology/ontology.go:152]: And: I'm not sure if I like to do operations in the var block already. But it is valid Go code so just my two cents

File: api/ontology/ontology.go [hunk 6]
```diff
@@ -142,3 +143,30 @@ func ResourceIDs(r []IsResource) []string {
...
+	)
+
+	for i := range od.Fields().Len() {
```
Comment (lebogg) [inline api/ontology/ontology.go:155]: I trust you on this. Maybe add some comment what we do here

File: api/ontology/ontology_test.go [hunk 7]
```diff
@@ -189,3 +188,36 @@ func TestListResourceIDs(t *testing.T) {
...
+	type args struct {
+		resource IsResource
+	}
+	tests := []struct {
+		name string
+		args args
+		want *Resource
+	}{
+		{
+			name: "happy path",
```
Comment (lebogg) [inline api/ontology/ontology_test.go:202]: Think about corner cases

File: service/assessment/assessment.go [hunk 8]
```diff
@@ -405,23 +404,11 @@ func (svc *Service) AssessEvidences(stream assessment.Assessment_AssessEvidences
...
-	}
-
-	err = api.Validate(m)
-	if err != nil {
-		return nil, err
-	}
-
-	resource, ok := m.(ontology.IsResource)
-	if !ok {
+	resource = ev.GetOntologyResource()
```
Comment (lebogg) [inline service/assessment/assessment.go:410]: much cleaner now :)

**Review by lebogg — COMMENTED** [review]
LGTM. Some minor things like adding comments and check if tests are covering the different scenarios

File: api/evidence/evidence.go [hunk 9]
```diff
@@ -41,43 +42,20 @@ func (req *StoreEvidenceRequest) GetPayload() proto.Message {
...
 		resource ontology.IsResource
-		err      error
 		ok       bool
 	)
-	// TODO: find a smarter way, because now we are unmarshalling the resource twice
-	// Try to extract the resource out of the evidence
-	m, err = ev.Resource.UnmarshalNew()
-	if err != nil {
+
+	if ev.Resource == nil || ev.Resource.Type == nil {
```
Comment (lebogg) [inline api/evidence/evidence.go:49]: Would it make sense to use `util.IsNil` here as well?

**Review by lebogg — APPROVED** [review]
LGTM

File: api/evidence/evidence.go (hunk 9 above)
Comment (oxisto) [reply api/evidence/evidence.go:49]: No, since ev.Resource is not an interface and we only need the utility function to check for nil on interfaces.
